### DISCUSSION TITLE
Inclusion of "holes" in surfaces and volumes

### DISCUSSION THREAD

**Discussion by blchoy** [team-sx/iwxxm-sx#3] [Ideas]
This is a point raised from Graeme Anderson who developed codes for the generation of IWXXM WAFS SIGWX F/C instances, and we observed similar expectation from ICAO WG-MOG who developed the new QVA service, that holes are inevitable in describing bounded surfaces and volumes.  Based on the form of information the MET community has conveyed so far we can safely assume that the features being described is within a bounded 2D horizontal area and 2.5D volume.  Having said that, we should confirm the necessity to mandate this constraint or otherwise software developer may be reluctant to invest on this capability as support for unnecessarily complex features is usually costly and error prone.

**Comment by blchoy** [comment]
Graeme further pointed out that:

_"Holes are something that we would probably want to include in objects once every few days, particularly for ICING and TURB objects. See the attached plot from a couple of weeks ago, where the gridded input data has thresholds applied so you can see where “moderate” (pink) and “severe” (yellow) turbulence occur. In the region around New Zealand where the jet splits you get a hole of lower turbulence. The lime green line indicates the bounds of the object created in this case, but there is clearly a very large area within this where turbulence is below the moderate threshold, so it would be more accurate to represent this with a hole in the polygon. I’ve also seen instances of turbulence wrapping the entire winter hemisphere pole, in which case the entire region covering the pole might need to be represented as a hole."_

![Turbulence Forecast](https://github.com/team-sx/iwxxm-sx/assets/30064374/ec1d2a85-9f51-42fc-851e-e0d83f87eb87)

**Comment by H-notle** [comment]
The usual structure/mechanism for this is that the points for shapes go one direction and holes go the other.
I think geojson & GIS, GML, KML (ISO 19125-1) have shapes being anti-clockwise/counter-clockwise while holes are clockwise. 
but that convention isn't universal as ESRI do the opposite.
is there any reason this approach is not possible for this problem?

**Comment by blchoy** [comment]
The original reason for putting this up, is to confirm the necessity of having holes in a polygon, as in our experience these holes are likely to be very tiny and can be safely ignored.  Without holes it will be simpler to implement the producing and consuming software.

Further discussion with Graeme indicated that holes are inevitable.  So we will be seeing this in upcoming WAFS SIGWX products.

Having said that, our 2.5D representation has a limitation, that the hole must go through the top and bottom of the horizontal slab.  If you have a hole which ends in the middle of the slab, you will have to divide this slab into two (one has a through hole, one has no hole).  This is not difficult if you have just one hole, but imagine you have piece of Swiss cheese with holes of various depths...

<img src="https://github.com/team-sx/iwxxm-sx/assets/30064374/ee9edc44-8b73-4ab8-8367-85b480752101">

**Comment by H-notle** [comment]
yes meteorology traditionally does draw features in the atmosphere like that - although I always thought that it was more minecraft shapes than emmental (always preferred gruyere myself) but its simplifying a complex shape. Its a compromise of  the fight between "keeping it simple" and "reasonably representative". But you could also have a shape and then in the metadata say that the tops/bases of the holes are only to FL200 or the equix. Your holey ones are only the hole versions of cloud shapes that used to be drawn as XXX/XXX   while the dimples specified the vertical extent(s) didn't reach both the vertical boundaries. 

But if you're dealing with the problem properly as data then really there should no need to simplify at all. So instead send shapes that just describe the pink and yellow in the above example and leave it to the downstream viewing tools to simplify if they want to

**Comment by blchoy** [comment]
> But if you're dealing with the problem properly as data then really there should no need to simplify at all. So instead send shapes that just describe the pink and yellow in the above example and leave it to the downstream viewing tools to simplify if they want to

This is a slab with holes:

```
<gml:patches>
    <gml:PolygonPatch>
        <gml:exterior>
            ... description of the polygon bounding your slab ...
        </gml:exterior>
        <gml:interior>
            ... description of your hole 1 ...
        </gml:interior>
        <gml:interior>
            ... description of your hole 2 ...
        </gml:interior>
    </gml:PolygonPatch>
</gml:patches>
<iwxxm:upperElevation .../>
<iwxxm:upperVerticalReference .../>
<iwxxm:lowerElevation .../>
<iwxxm:lowerVerticalReference .../>
```

>But you could also have a shape and then in the metadata say that the tops/bases of the holes are only to FL200 or the equix. Your holey ones are only the hole versions of cloud shapes that used to be drawn as XXX/XXX while the dimples specified the vertical extent(s) didn't reach both the vertical boundaries.

This is a slab with no hole and separate indication of holes with various thickness:

```
<gml:patches>
    <gml:PolygonPatch>
        <gml:exterior>
            ... description of the polygon bounding your slab...
        </gml:exterior>
    </gml:PolygonPatch>
</gml:patches>
<iwxxm:upperElevation .../>
<iwxxm:upperVerticalReference .../>
<iwxxm:lowerElevation .../>
<iwxxm:lowerVerticalReference .../>
...
<gml:patches>
    <gml:PolygonPatch>
        <gml:exterior>
            ... description of the polygon bounding your hole 1...
        </gml:exterior>
    </gml:PolygonPatch>
</gml:patches>
<iwxxm:upperElevation .../>
<iwxxm:upperVerticalReference .../>
<iwxxm:lowerElevation .../>
<iwxxm:lowerVerticalReference .../>
...
<gml:patches>
    <gml:PolygonPatch>
        <gml:exterior>
            ... description of the polygon bounding your hole 2...
        </gml:exterior>
    </gml:PolygonPatch>
</gml:patches>
<iwxxm:upperElevation .../>
<iwxxm:upperVerticalReference .../>
<iwxxm:lowerElevation .../>
<iwxxm:lowerVerticalReference .../>
```

Now the second representation solves the problem of holes with various depths, but in this case as you mentioned it will be some metadata saying they are holes, instead of the orientation of the bounding polygon.  If this is the enhanced representation for SIGWX F/C we will need to revise the schema.

**Comment by blchoy** [comment]
Closed because it has been dormant for quite some time.
